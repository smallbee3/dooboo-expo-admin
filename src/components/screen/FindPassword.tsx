import { Button, EditText } from 'dooboo-ui';
import {
  FindPasswordMutation,
  FindPasswordMutationResponse,
} from '../../__generated__/FindPasswordMutation.graphql';
import React, { ReactElement, useState } from 'react';
import { graphql, useMutation } from 'react-relay/hooks';

import { AuthStackNavigationProps } from '../navigation/AuthNavigator';
import { getString } from '../../../STRINGS';
import styled from 'styled-components/native';
import { useThemeContext } from '@dooboo-ui/theme';
import { validateEmail } from '../../utils/common';

interface Props {
  navigation: AuthStackNavigationProps<'FindPassword'>;
}

const Container = styled.View`
  flex: 1;
  justify-content: center;
  align-items: center;
  background-color: ${({ theme }): string => theme.background};
`;

const ContentWrapper = styled.View`
  width: 528px;
`;

const Title = styled.Text`
  font-style: normal;
  font-weight: bold;
  font-size: 36px;
  line-height: 53px;

  color: ${({ theme }): string => theme.font};

  width: 100%;
  text-align: center;
  margin-bottom: 18px;
`;

const SubTitle = styled.Text`
  font-style: normal;
  font-weight: normal;
  font-size: 24px;
  line-height: 36px;

  align-items: center;
  text-align: center;
  margin-bottom: 59px;

  color: ${({ theme }): string => theme.font};
`;

const signUp = graphql`
  mutation FindPasswordMutation($email: String!) {
    findPassword(email: $email)
  }
`;

const ResetPassword = (props: Props): ReactElement => {
  const { theme } = useThemeContext();
  const [email, setEmail] = useState<string>('');
  const [error, setError] = useState<string>('');
  const [commitSignUp, isInFlight] = useMutation<FindPasswordMutation>(signUp);
  const mutationConfig = {
    variables: {
      email,
    },
    onCompleted: (response: FindPasswordMutationResponse): void => {
      if (response.findPassword) {
        return props.navigation.navigate('ConfirmEmail');
      }
    },
    onError: (error: any): void => {
      console.error(error);
      setError(getString('ALERT_FAILED_FIND_PASSWORD'));
    },
  };

  const handleChangeEmail = (text: string): void => {
    if (text === '') setError('');
    setEmail(text);
  };

  const handleSubmit = (): void => {
    if (!validateEmail(email)) {
      return setError(getString('ALERT_FAILED_FIND_PASSWORD'));
    }

    setError('');

    commitSignUp(mutationConfig);
  };
  return (
    <Container>
      <ContentWrapper>
        <Title>{getString('RESET_PASSWORD')}</Title>
        <SubTitle>{getString('ENTER_EMAIL_FOR_TEMPORARY_PASSWORD')}</SubTitle>
        <EditText
          testID="input-email"
          style={{
            marginBottom: 65,
          }}
          textStyle={{
            fontSize: 16,
            lineHeight: 24,
            alignItems: 'center',
            color: theme.font,
          }}
          value={email}
          onChangeText={handleChangeEmail}
          placeholder={getString('EMAIL')}
          label={getString('EMAIL')}
          labelTextStyle={{
            fontWeight: 'bold',
            fontSize: 14,
            lineHeight: 21,
          }}
          errorText={error}
        />
        <Button
          testID="btn-submit"
          style={{
            backgroundColor: theme.btnPrimary,
            borderRadius: 6,
            width: '100%',
            height: 48,
            alignItems: 'center',
          }}
          textStyle={{
            fontSize: 14,
            fontWeight: 'bold',
            color: theme.btnPrimaryFont,
            lineHeight: 21,
            letterSpacing: -0.5,
          }}
          onPress={handleSubmit}
          text={getString('VERIFY_EMAIL')}
          isLoading={isInFlight}
        />
      </ContentWrapper>
    </Container>
  );
};

export default ResetPassword;
