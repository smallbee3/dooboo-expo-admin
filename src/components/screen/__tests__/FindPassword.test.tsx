import 'react-native';

import React, { ReactElement } from 'react';
import {
  RenderResult,
  cleanup,
  fireEvent,
  render,
  waitForElement,
} from '@testing-library/react-native';
import {
  createTestElement,
  createTestProps,
  environment,
} from '../../../../test/testUtils';

import FindPassword from '../FindPassword';
import { MockPayloadGenerator } from 'relay-test-utils';
import { act } from 'react-test-renderer';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
let props: any;
let component: ReactElement;
let testingLib: RenderResult;

describe('[FindPassword] Rendering', () => {
  beforeEach(() => {
    props = createTestProps();
    component = createTestElement(<FindPassword {...props} />);
    testingLib = render(component);
  });

  it('renders without crashing', () => {
    const { baseElement } = testingLib;
    expect(baseElement).toMatchSnapshot();
    expect(baseElement).toBeTruthy();
  });
});

// describe('[FindPassword] Interaction', () => {
//   beforeEach(() => {
//     props = createTestProps();
//     component = createTestElement(<FindPassword {...props} />);
//     testingLib = render(component);
//   });

//   it('should invoke changeText when email changed', async () => {
//     const emailText = testingLib.getByTestId('input-email');
//     await waitForElement(() => emailText);

//     act(() => {
//       fireEvent.changeText(emailText, 'email@email.com');
//     });

//     expect(emailText.props.value).toEqual('email@email.com');
//   });

//   it('should make request to findPassword when button has pressed', async () => {
//     const emailText = testingLib.getByTestId('input-email');
//     await waitForElement(() => emailText);

//     act(() => {
//       fireEvent.changeText(emailText, 'email@email.com');
//     });

//     const submitBtn = testingLib.getByTestId('btn-submit');
//     await waitForElement(() => submitBtn);

//     act(() => {
//       fireEvent.press(submitBtn);

//       environment.mock.resolveMostRecentOperation((operation) => {
//         MockPayloadGenerator.generate(operation);
//       });
//     });
//   });

//   afterAll((done) => {
//     cleanup();
//     done();
//   });
// });
