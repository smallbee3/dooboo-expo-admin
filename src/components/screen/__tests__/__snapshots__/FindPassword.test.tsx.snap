// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[FindPassword] Rendering renders without crashing 1`] = `
<View
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    collapsable={true}
    pointerEvents="box-none"
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "white",
            "flexBasis": 0,
            "flexGrow": 1,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "width": 528,
            },
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#1B1B21",
                "fontSize": 36,
                "fontStyle": "normal",
                "fontWeight": "bold",
                "lineHeight": 53,
                "marginBottom": 18,
                "textAlign": "center",
                "width": "100%",
              },
            ]
          }
        >
          Reset password
        </Text>
        <Text
          style={
            Array [
              Object {
                "alignItems": "center",
                "color": "#1B1B21",
                "fontSize": 24,
                "fontStyle": "normal",
                "fontWeight": "normal",
                "lineHeight": 36,
                "marginBottom": 59,
                "textAlign": "center",
              },
            ]
          }
        >
          Please enter an email address to issue a temporary password
        </Text>
        <View
          style={
            Array [
              Object {
                "alignSelf": "stretch",
                "display": "flex",
                "flexDirection": "column",
              },
              Object {
                "marginBottom": 65,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#b9b9c4",
                  "fontSize": 14,
                  "fontWeight": "500",
                  "marginBottom": 5,
                },
                Object {
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "lineHeight": 21,
                },
                null,
              ]
            }
          >
            Email
          </Text>
          <TextInput
            allowFontScaling={true}
            autoCapitalize="none"
            multiline={false}
            placeholder="Email"
            rejectResponderTermination={true}
            style={
              Array [
                Object {
                  "fontSize": 15,
                  "fontWeight": "500",
                  "paddingBottom": 15,
                  "paddingTop": 15,
                },
                Object {
                  "alignItems": "center",
                  "color": "#1B1B21",
                  "fontSize": 16,
                  "lineHeight": 24,
                },
              ]
            }
            testID="input-email"
            underlineColorAndroid="transparent"
            value=""
          />
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "width": "100%",
                },
                undefined,
                Object {
                  "borderBottomWidth": 0.6,
                  "borderColor": "#eaeaf9",
                },
                null,
              ]
            }
          />
        </View>
        <View
          delayPressIn={30}
          testID="btn-submit"
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "alignSelf": "center",
                  "borderColor": "blue",
                  "flexDirection": "row",
                  "height": 52,
                  "justifyContent": "center",
                  "width": 320,
                },
                Object {
                  "alignItems": "center",
                  "backgroundColor": "#609FFF",
                  "borderRadius": 6,
                  "height": 48,
                  "width": "100%",
                },
              ]
            }
          >
            <Text
              style={
                Array [
                  Object {
                    "color": "#069ccd",
                    "fontSize": 16,
                  },
                  Object {
                    "color": "white",
                    "fontSize": 14,
                    "fontWeight": "bold",
                    "letterSpacing": -0.5,
                    "lineHeight": 21,
                  },
                ]
              }
            >
              Verify email
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
